### PR TITLE
Scmplus

### PR DESCRIPTION
**Pull request by evilpete**
SCM+ decoder

written in response to #1378

tested only on samples as I do not have a local source to test with (my meter uses something different )

### CODE REVIEW

**Comment by evilpete** [comment]
@merbanan   :  I named this  scmplus, but it occurred to me that the (regular) scm decoder you already have is named [ert](/merbanan/rtl_433/blob/master/src/devices/ert.c),  is there preferred name for SCM+ for continuity ?

**Comment by evilpete** [comment]
Note: 
Field key names and json format set to mostly match [rtlamr](https://github.com/bemasher/rtlamr) field names.

A [pull request](/merbanan/rtl_433_tests/pull/347) has been made to rtl_433_tests with samples

File: src/devices/scmplus.c [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+    Copyright (C) 2020 Peter Shipley <peter.shipley@gmail.com>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+*/
+
+#include <arpa/inet.h>
```
Comment (merbanan) [inline src/devices/scmplus.c:12]: Can we get rid of this ?

File: src/devices/scmplus.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    /* clang-format off */
+    data = data_make(
+            "model",           "",                 DATA_STRING, "SCM+",
+            "ProtocolID",     "Protocol_ID",       DATA_STRING, protocol_id_str,
+            "EndpointType",   "Endpoint_Type",     DATA_STRING, endpoint_type_str,
+            "EndpointID",     "Endpoint_ID",       DATA_INT,    endpoint_id,
+            "Consumption",     "",                 DATA_INT,    consumption_data,
+            "Tamper",          "",                 DATA_STRING, physical_tamper_str,
+            "PacketCRC",       "crc",              DATA_STRING, crc_str,
```
Comment (merbanan) [inline src/devices/scmplus.c:161]: This value is not really needed but can be kept for rtlamr compatibility reasons.

File: src/devices/scmplus.c [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    /* 
+        Field key names and format set to  match rtlamr field names 
+
+        {Time:2020-06-20T09:58:19.074 Offset:49152 Length:49152
+        SCM+:{ProtocolID:0x1E EndpointType:0xAB EndpointID:  68211547 Consumption:  6883 Tamper:0x4900 PacketCRC:0x39BE}}
+    */
+
+    /* clang-format off */
+    data = data_make(
+            "model",           "",                 DATA_STRING, "SCM+",
```
Comment (merbanan) [inline src/devices/scmplus.c:155]: The + character might not be such a good mqtt topic. I'm not sure but if so the a change from + to plus would be good.

File: src/devices/scmplus.c [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+/**
+
+Freq 912600155
+
+Random information:
+
+https://github.com/bemasher/rtlamr/wiki/Protocol
+http://www.gridinsight.com/community/documentation/itron-ert-technology/
+
+ Units: Some meter types transmit consumption in 1 kWh units, while others use more granular 10 Wh units
```
Comment (merbanan) [inline src/devices/scmplus.c:24]: Is there no way to know this ?

File: src/devices/scmplus.c [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    if (sync_index >= bitbuffer->bits_per_row[0]) {
+        return DECODE_ABORT_EARLY;
+    }
+
+    if ( (bitbuffer->bits_per_row[0] - sync_index) < 128) {
+        return DECODE_ABORT_LENGTH;
+    }
+
+    if (decoder->verbose) {
+        (void)fprintf(stderr, "%s: row len=%hu\n", __func__, bitbuffer->bits_per_row[0]);
```
Comment (merbanan) [inline src/devices/scmplus.c:50]: void should not be needed

File: src/devices/scmplus.c (hunk 4 above)
Comment (evilpete) [reply src/devices/scmplus.c:24]: not that i can find.  rtlamr has the same comment

File: src/devices/scmplus.c (hunk 3 above)
Comment (stevenjev) [reply src/devices/scmplus.c:155]: Testing this pull request, when using [/model], I received "SCM_" not "SCM+"

Comment (merbanan) [reply src/devices/scmplus.c:155]: SCMplus is prefered then.

File: src/devices/scmplus.c (hunk 1 above)
Comment (evilpete) [reply src/devices/scmplus.c:12]: your right I should have used `htobe16` from  `endian.h `?   
although I find `ntohs()` to be more portable.

I would rather not automatically make assumptions on local byte order with  hard a coded
`((t_16 & 0xFF00) >> 8 | (t_16 & 0x00FF) << 8)`

with intermixed with :
`#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__`

File: src/devices/scmplus.c [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        (void)fprintf(stderr, "%s: sync_index=%d\n", __func__, sync_index);
+    }
+
+    // bitbuffer_debug(bitbuffer);
+    bitbuffer_extract_bytes(bitbuffer, 0, sync_index, b, 16 * 8);
+
+    if (decoder->verbose) { // print bytes with aligned offset
```
[the remark this answers is not in the document]
Comment (evilpete) [reply src/devices/scmplus.c:57]: done

File: src/devices/scmplus.c (hunk 2 above)
Comment (evilpete) [reply src/devices/scmplus.c:161]: yes it was included for theoretical  compatibility.

compatibility is also the reason I convert the fields to hex strings manually as opposed to using DATA_FORMAT

File: src/devices/scmplus.c (hunk 3 above)
Comment (evilpete) [reply src/devices/scmplus.c:155]: done

File: src/devices/scmplus.c (hunk 5 above)
Comment (evilpete) [reply src/devices/scmplus.c:50]: Sorry, old-school habit

File: src/devices/scmplus.c (hunk 1 above)
Comment (zuckschwerdt) [reply src/devices/scmplus.c:12]: I'd say we are used to pick bit and bytes from random places in the decoders, for just two instances of short with not real performance requirement just pick t_16 apart and just bitshift-or.

Comment (evilpete) [reply src/devices/scmplus.c]: ok...
bitshift-or  manually 
Done

**Pull request merged**